pytensor/tensor/blockwise.py

```diff
@@ -0,0 +1,431 @@
+        signature
+            Generalized universal function signature,
+            e.g., (m,n),(n)->(m) for vectorized matrix-vector multiplication
+
+        """
+        if isinstance(core_op, Blockwise):
+            raise TypeError("Core Op is already a Blockwise")
+
+        if signature is None:
+            signature = getattr(core_op, "gufunc_signature", None)
+            if signature is None:
+                raise ValueError(
+                    f"Signature not provided nor found in core_op {core_op}"
+                )
+
+        self.core_op = core_op
+        self.signature = signature
+        self.name = name
+        self.inputs_sig, self.outputs_sig = _parse_gufunc_signature(signature)
+        self._gufunc = None
+        super().__init__(**kwargs)
+
+    def _create_dummy_core_node(self, inputs: Sequence[TensorVariable]) -> Apply:
+        core_input_types = []
+        for i, (inp, sig) in enumerate(zip(inputs, self.inputs_sig)):
+            if inp.type.ndim < len(sig):
+                raise ValueError(
+                    f"Input {i} {inp} has insufficient core dimensions for signature {self.signature}"
+                )
+            # ndim_supp = 0 case
+            if not sig:
+                core_shape = ()
+            else:
+                core_shape = inp.type.shape[-len(sig) :]
+            core_input_types.append(tensor(dtype=inp.type.dtype, shape=core_shape))
+
+        core_node = self.core_op.make_node(*core_input_types)
+
+        if len(core_node.outputs) != len(self.outputs_sig):
+            raise ValueError(
+                f"Insufficient number of outputs for signature {self.signature}: {len(core_node.outputs)}"
+            )
+        for i, (core_out, sig) in enumerate(zip(core_node.outputs, self.outputs_sig)):
+            if core_out.type.ndim != len(sig):
+                raise ValueError(
+                    f"Output {i} of {self.core_op} has wrong number of core dimensions for signature {self.signature}: {core_out.type.ndim}"
+                )
+
+        return core_node
+
+    def make_node(self, *inputs):
+        inputs = [as_tensor_variable(i) for i in inputs]
+
+        core_node = self._create_dummy_core_node(inputs)
+
+        batch_ndims = max(
+            inp.type.ndim - len(sig) for inp, sig in zip(inputs, self.inputs_sig)
+        )
+
+        batched_inputs = []
+        batch_shapes = []
+        for i, (inp, sig) in enumerate(zip(inputs, self.inputs_sig)):
+            # Append missing dims to the left
+            missing_batch_ndims = batch_ndims - (inp.type.ndim - len(sig))
+            if missing_batch_ndims:
+                inp = shape_padleft(inp, missing_batch_ndims)
+            batched_inputs.append(inp)
+
+            if not sig:
+                batch_shapes.append(inp.type.shape)
+            else:
+                batch_shapes.append(inp.type.shape[: -len(sig)])
+
+        def get_most_specialized_batch_shape(
```

Comment:
I remember I saw similar code in elemwise, can we avoid duplication there?

Reply:
refactored